### PR TITLE
Problem: missing high-level docs for x/cronos

### PR DESCRIPTION
**Pull request by yihuang**
Closes: #64

Solution:
- write cronos module spec

👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest master? 
- [ ] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

File: x/cronos/spec/01_concepts.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+Native token is a token managed by cosmos native bank module, there are several kinds of native tokens in Cronos:
+
+- gas token. used to pay the gas fee.
+- staking token. used to do PoA consensus.
+- IBC token. tokens come from IBC channels.
+- gravity token. tokens come from the gravity bridge.
+
+## CRC20 Token
+
+CRC20 token is Cronos's equivalence of ERC20 token on Ethereum with some extensions, they can be mapped with native tokens and support transfer to/from native tokens, and potentially transfer to/from Ethereum and another cosmos chain through gravity bridge and IBC.
```
Comment (thomas-nguy) [inline x/cronos/spec/01_concepts.md:33]: I think the short description `CRC20 token is Cronos's equivalence of ERC20 token on Ethereum with some extensions` is enough.

We probably want to have more details in another sections for CRC proposal

File: x/cronos/spec/02_state.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+|                         | Key                                    | Value                      |
+| ----------------------- | -------------------------------------- | -------------------------- |
+| DenomToExternalContract | `[]byte{1} + []byte(denom)`            | `[]byte(contract_address)` |
+| DenomToAutoContract     | `[]byte{2} + []byte(denom)`            | `[]byte(contract_address)` |
+| ContractToDenom         | `[]byte{3} + []byte(contract_address)` | `[]byte(denom)`            |
+
+- `DenomToExternalContract` stores a map from denom to external CRC20 contract.
+- `DenomToAutoContract` stores a map from denom to auto-deployed CRC20 contract.
+- `ContractToDenom` stored the reversed map for both external and auto-deployed contracts.
```
Comment (aw126) [inline x/cronos/spec/02_state.md:17]:
```suggestion
- `ContractToDenom` stores the reversed map for both external and auto-deployed contracts.
```

File: x/cronos/spec/04_messages.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+## MsgConvertVouchers
+
+Convert native tokens to the mapped CRC20 tokens, if the mapping does not exist and auto-deployment is enabled, an embed CRC20 contract is deployed for it automatically, otherwise, the message fails.
+
++++ https://github.com/crypto-org-chain/cronos/blob/v0.6.0-testnet/proto/cronos/tx.proto#L26-L30
+
+This message is expected to fail if:
+
+- The coin denom is neither IBC nor gravity tokens.
+- the mapping does not exist and auto-deployment is not enabled.
```
Comment (aw126) [inline x/cronos/spec/04_messages.md:14]:
```suggestion
- The mapping does not exist and auto-deployment is not enabled.
```

File: x/cronos/spec/05_abci.md [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+## InitGenesis
+
+`InitGenesis` initializes the Cronos module genesis state by setting the `GenesisState` fields to the
+store. In particular it sets the parameters and token mapping state.
+
+## ExportGenesis
+
+The `ExportGenesis` ABCI function exports the genesis state of the Cronos module. In particular, it
+iterate all token mappings to genesis.
```
Comment (aw126) [inline x/cronos/spec/05_abci.md:15]:
```suggestion
iterates all token mappings to genesis.
```

File: x/cronos/spec/README.md [hunk 5]
```diff
@@ -0,0 +1,22 @@
+<!--
+order: 0
+title: EVM Overview
+parent:
+  title: "evm"
```
[the remark this answers is not in the document]
Comment (yihuang) [reply x/cronos/spec/README.md]: good catch, fixed.

File: x/cronos/spec/03_state_transitions.md [hunk 6]
```diff
@@ -0,0 +1,7 @@
+<!--
+order: 3
+-->
+
+# State Transitions
+
+<!-- define state transitions for accounts and storage -->
```
[the remark this answers is not in the document]
Comment (yihuang) [reply x/cronos/spec/03_state_transitions.md]: added sth about the CRUD of token mapping.

File: x/cronos/spec/01_concepts.md (hunk 1 above)
Comment (yihuang) [reply x/cronos/spec/01_concepts.md:33]: The CRC standard can be described in another place, not necessarily a part of Cronos module spec I think.

File: x/cronos/spec/01_concepts.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+The CROs on Cronos chain need to transfered from Crypto.org chain through IBC channel, they arrive in Cronos chain as IBC tokens first, then get converted to gas token automatically.
+
+### Decimal Places Conversion
+
+The CRO on Crypto.org chain has 8 decimals, while Cronos gas token has 18 decimals (to keep compatibility with Ethereum). So a convertion need to be done:
+
+- When transfer CRO to Cronos chain, the decimal places of the amount are expanded to 18.
+- When transfer CRO from Cronos chain, the amount are truncated to 8 decimals, the remaining part is left in Cronos, so the total value is preserved.
+
```
[the remark this answers is not in the document]
Comment (yihuang) [reply x/cronos/spec/01_concepts.md:21]: there's a specific doc to cover the messages.

File: x/cronos/spec/01_concepts.md [hunk 8]
```diff
@@ -0,0 +1,45 @@
+ <!-- order: 1 -->
+
+# Concepts
+
+## Cronos
+
+Cronos is the name of Crypto.org EVM Chain, also the name of the main module that glues multiple components together for Cronos chain.
```
[the remark this answers is not in the document]
Comment (yihuang) [reply x/cronos/spec/01_concepts.md]: done

File: x/cronos/spec/01_concepts.md [hunk 9]
```diff
@@ -0,0 +1,49 @@
+ <!-- order: 1 -->
+
+# Concepts
+
+## Cronos
+
+Cronos is an EVM chain in Crypto.org ecosystem. It allows users to deploy and interact with smart contracts. Powered by Ethermint, it is built using cosmos SDK which allows to leverage to full capability of the cosmos ecosystem. It is also connected to the ethereum network with the gravity bridge integration (WIP).
+
+Bridging asset from cosmos or ethereum are automatically converted to a CRC20 asset when moved to Cronos which make it extremely easy to integrate with existing web3 tools.
```
Comment (tomtau) [inline x/cronos/spec/01_concepts.md:9]: maybe a link to CRC20 definition?

File: x/cronos/spec/01_concepts.md [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+The Cronos module glues IBC, gravity bridge and ethermint together through hooks and token mapping.
+
+## Gas Token
+
+Cronos use CRO as the gas token.
+
+The CROs on the Cronos chain need to be transferred from the Crypto.org chain through an IBC channel. The tokens arrive on the Cronos chain as IBC tokens first, then automatically convert to the gas token.
```
Comment (tomtau) [inline x/cronos/spec/01_concepts.md:17]:
```suggestion
The CRO assets on the Cronos Chain need to be transferred from the Crypto.org Chain through an IBC channel. The tokens arrive on the Cronos Chain as IBC tokens first, then are automatically converted to the gas token. 
```

File: x/cronos/spec/01_concepts.md [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+Cronos is an EVM chain in Crypto.org ecosystem. It allows users to deploy and interact with smart contracts. Powered by Ethermint, it is built using cosmos SDK which allows to leverage to full capability of the cosmos ecosystem. It is also connected to the ethereum network with the gravity bridge integration (WIP).
+
+Bridging asset from cosmos or ethereum are automatically converted to a CRC20 asset when moved to Cronos which make it extremely easy to integrate with existing web3 tools.
+
+The Cronos module glues IBC, gravity bridge and ethermint together through hooks and token mapping.
+
+## Gas Token
+
+Cronos use CRO as the gas token.
```
Comment (tomtau) [inline x/cronos/spec/01_concepts.md:15]:
```suggestion
Cronos uses CRO as its gas token.
```

File: x/cronos/spec/01_concepts.md (hunk 9 above)
Comment (tomtau) [reply x/cronos/spec/01_concepts.md:9]: oh, it's briefly described below -- but may be good to link the actual crc20 doc describing the interface

File: x/cronos/spec/01_concepts.md [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+- IBC token. tokens come from IBC channels.
+- gravity token. tokens come from the gravity bridge.
+
+## CRC20 Token
+
+CRC20 token is Cronos's equivalence of ERC20 token on Ethereum with some extensions, they can be mapped with native tokens and support transfer to/from native tokens, and potentially transfer to/from Ethereum and another cosmos chain through gravity bridge and IBC.
+
+## Auto-deployed Contract
+
+A contract whose byte code is embedded in Cronos module and deployed by it, and some operations in it are only authorized to Cronos module. These contracts can be trusted by Cronos module directly. Currently, Cronos module support auto-deploy a minimal CRC20 contract to support automatic wrapping native token in EVM.
```
Comment (tomtau) [inline x/cronos/spec/01_concepts.md:41]: maybe linking the contract description or sketching out its code?

File: x/cronos/spec/07_params.md [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+| `KeyIbcTimeout`           | uint64 | `86400000000000`                                             |
+| `KeyCronosAdmin`          | string | `""`                                                         |
+| `KeyEnableAutoDeployment` | bool   | `false`                                                      |
+
+- `IbcCroDenom` Specifies the IBC token that should be converted to gas token upon arrival automatically.
+
+- `KeyIbcTimeout` The timeout value to use when Cronos module transfer tokens to IBC channels.
+
+- `KeyCronosAdmin` The account that is authorized to manage token mapping through message, empty means no admin, should be a valid bech32 cosmos address if specified.
+- `KeyEnableAutoDeployment` specifies if the auto-deployment feature is enabled.
```
Comment (tomtau) [inline x/cronos/spec/07_params.md:21]: what happens when auto-deployment is off and then it's turned on via paramchange proposal? maybe good to add some comments on these parameters whether they are easily "updatable" -- to me, it seems only CronosAdmin can be updated via paramchange... the others look like that they'd need some extra code to be executed for store migrations etc. (so perhaps there should be a warning that they are not easily updatable, but would need some custom upgrade logic?)

File: x/cronos/spec/07_params.md [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+| `IbcCroDenom`             | string | `"ibc/6B5A664BF0AF4F71B2F0BAA33141E2F1321242FBD5D19762F541EC971ACB0865"` |
+| `KeyIbcTimeout`           | uint64 | `86400000000000`                                             |
+| `KeyCronosAdmin`          | string | `""`                                                         |
+| `KeyEnableAutoDeployment` | bool   | `false`                                                      |
+
+- `IbcCroDenom` Specifies the IBC token that should be converted to gas token upon arrival automatically.
+
+- `KeyIbcTimeout` The timeout value to use when Cronos module transfer tokens to IBC channels.
+
+- `KeyCronosAdmin` The account that is authorized to manage token mapping through message, empty means no admin, should be a valid bech32 cosmos address if specified.
```
Comment (tomtau) [inline x/cronos/spec/07_params.md:20]: https://github.com/crypto-org-chain/cronos/pull/185 also "Key" to be removed

File: x/cronos/spec/04_messages.md [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
++++ https://github.com/crypto-org-chain/cronos/blob/v0.6.0-testnet/proto/cronos/tx.proto#L33-L38
+
+This message is expected to fail if:
+
+- The sender doesn't have enough balance.
+- The IBC transfer message fails.
+
+## MsgUpdateTokenMapping
+
+Update external token mapping, insert if not exists, can only be called by Cronos admin account, which is configured in module parameters.
```
Comment (tomtau) [inline x/cronos/spec/04_messages.md:29]: how is the supply managed if there was an existing token mapping, and it was updated to a different one?

File: x/cronos/spec/04_messages.md [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
++++ https://github.com/crypto-org-chain/cronos/blob/v0.6.0-testnet/proto/cronos/tx.proto#L26-L30
+
+This message is expected to fail if:
+
+- The coin denom is neither IBC nor gravity tokens.
+- The mapping does not exist and auto-deployment is not enabled.
+
+## MsgTransferTokens
+
+Transfer IBC tokens away from Cronos chain, will do decimals conversion automatically for CRO.
```
Comment (tomtau) [inline x/cronos/spec/04_messages.md:18]: maybe some details on what it does -- I assume it'll do the conversion and do some implicit action with ibc-transfer: https://github.com/cosmos/ibc-go/blob/main/proto/ibc/applications/transfer/v1/tx.proto#L20 with some of those fields hardcoded?

File: x/cronos/spec/04_messages.md [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+Convert native tokens to the mapped CRC20 tokens, if the mapping does not exist and auto-deployment is enabled, an embed CRC20 contract is deployed for it automatically, otherwise, the message fails.
+
++++ https://github.com/crypto-org-chain/cronos/blob/v0.6.0-testnet/proto/cronos/tx.proto#L26-L30
+
+This message is expected to fail if:
+
+- The coin denom is neither IBC nor gravity tokens.
+- The mapping does not exist and auto-deployment is not enabled.
+
+## MsgTransferTokens
```
Comment (tomtau) [inline x/cronos/spec/04_messages.md:16]: is this the reverse of `MsgConvertVouchers` or how does one "exit" the EVM contract? what if the contract locks the tokens in some way?

File: x/cronos/spec/04_messages.md [hunk 18]
```diff
@@ -0,0 +1,36 @@
+<!-- order: 4 -->
+
+# Messages
```
Comment (tomtau) [inline x/cronos/spec/04_messages.md:3]: one useful information for `MsgConvertVouchers` and `MsgTransferTokens` would be to specify from/to addresses which ecosystem they are expected to come from... I assume `address` in `MsgConvertVouchers` would the be used in EVM, so it should come from the Ethereum ecosystem, but when on first transfers in the IBC tokens, they may land on an address that comes from the Cosmos ecosystem?

or would those message check the account types?

File: x/cronos/spec/03_state_transitions.md [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+
+This document describes the state transition operations of Cronos module, currently the only state is token mapping.
+
+## Token Mapping
+
+### Create/Update
```
Comment (tomtau) [inline x/cronos/spec/03_state_transitions.md:13]: how are the supplies transitioned on updates?

**Pull request merged**